samples/bluetooth/channel_sounding/src/distance_estimation.c

```diff
@@ -0,0 +1,303 @@
+	enum bt_conn_le_cs_role role;
+};
+
+static void calc_complex_product(int32_t z_a_real, int32_t z_a_imag, int32_t z_b_real,
+				 int32_t z_b_imag, int32_t *z_out_real, int32_t *z_out_imag)
+{
+	*z_out_real = z_a_real * z_b_real - z_a_imag * z_b_imag;
+	*z_out_imag = z_a_real * z_b_imag + z_a_imag * z_b_real;
+}
+
+static double linear_regression(double *x_values, double *y_values, uint8_t n_samples)
+{
+	if (n_samples == 0) {
+		return 0.0;
+	}
+
+	/* Estimates b in y = a + b x */
+
+	double y_mean = 0.0;
+	double x_mean = 0.0;
+
+	for (uint8_t i = 0; i < n_samples; i++) {
+		y_mean += (y_values[i] - y_mean) / (i + 1);
+		x_mean += (x_values[i] - x_mean) / (i + 1);
+	}
+
+	double b_est_upper = 0.0;
+	double b_est_lower = 0.0;
+
+	for (uint8_t i = 0; i < n_samples; i++) {
+		b_est_upper += (x_values[i] - x_mean) * (y_values[i] - y_mean);
+		b_est_lower += (x_values[i] - x_mean) * (x_values[i] - x_mean);
+	}
+
+	return b_est_upper / b_est_lower;
+}
+
+static void bubblesort_2(double *array1, double *array2, uint16_t len)
+{
+	bool swapped;
+	double temp;
+
+	for (uint16_t i = 0; i < len - 1; i++) {
+		swapped = false;
+		for (uint16_t j = 0; j < len - i - 1; j++) {
+			if (array1[j] > array1[j + 1]) {
+				temp = array1[j];
+				array1[j] = array1[j + 1];
+				array1[j + 1] = temp;
+				temp = array2[j];
+				array2[j] = array2[j + 1];
+				array2[j + 1] = temp;
+				swapped = true;
+			}
+		}
+
+		if (!swapped) {
+			break;
+		}
+	}
+}
+
+static double estimate_distance_using_phase_slope(struct iq_sample_and_channel *data, uint8_t len)
+{
+	int32_t combined_i;
+	int32_t combined_q;
+	uint16_t num_angles = 0;
+	static double theta[MAX_NUM_SAMPLES];
+	static double frequencies[MAX_NUM_SAMPLES];
+
+	for (uint8_t i = 0; i < len; i++) {
+		if (!data[i].failed) {
+			calc_complex_product(data[i].local_iq_sample.i, data[i].local_iq_sample.q,
+					     data[i].peer_iq_sample.i, data[i].peer_iq_sample.q,
+					     &combined_i, &combined_q);
+
+			theta[num_angles] = atan2(1.0 * combined_q, 1.0 * combined_i);
+			frequencies[num_angles] = 1.0 * CS_FREQUENCY_MHZ(data[i].channel);
+			num_angles++;
+		}
+	}
+
+	if (num_angles < 2) {
+		return 0.0;
+	}
+
+	/* Sort phases by tone frequency */
+	bubblesort_2(frequencies, theta, num_angles);
+
+	/* One-dimensional phase unwrapping */
+	for (uint8_t i = 1; i < num_angles; i++) {
+		double difference = theta[i] - theta[i - 1];
+
+		if (difference > PI) {
+			for (uint8_t j = i; j < num_angles; j++) {
+				theta[j] -= 2.0 * PI;
+			}
+		} else if (difference < -PI) {
+			for (uint8_t j = i; j < num_angles; j++) {
+				theta[j] += 2.0 * PI;
+			}
+		}
+	}
+
+	double phase_slope = linear_regression(frequencies, theta, num_angles);
+
+	double distance = -phase_slope * (SPEED_OF_LIGHT_M_PER_S / (4 * PI));
+
+	return distance / 1000000.0; /* Scale to meters. */
+}
+
+static double estimate_distance_using_time_of_flight(uint8_t n_samples)
+{
+	double tof;
+	double tof_mean = 0.0;
+
+	/* Cumulative Moving Average */
```

Comment:
Is there a reason to do this a as moving average and not just an average?

Reply:
I think this code arrives at the same result. I wanted to avoid calculating a big sum to avoid potential overflows. But, to be honest, the range of a float is big enough that this probably doesn't matter.

The approach shown here might be more appropriate with fixed-point variables